server/deviceflowhandlers.go

```diff
@@ -0,0 +1,386 @@
+package server
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/dexidp/dex/storage"
+)
+
+type deviceCodeResponse struct {
+	//The unique device code for device authentication
+	DeviceCode string `json:"device_code"`
+	//The code the user will exchange via a browser and log in
+	UserCode string `json:"user_code"`
+	//The url to verify the user code.
+	VerificationURI string `json:"verification_uri"`
+	//The verification uri with the user code appended for pre-filling form
+	VerificationURIComplete string `json:"verification_uri_complete"`
+	//The lifetime of the device code
+	ExpireTime int `json:"expires_in"`
+	//How often the device is allowed to poll to verify that the user login occurred
+	PollInterval int `json:"interval"`
+}
+
+func (s *Server) getDeviceVerificationURI() string {
+	return path.Join(s.issuerURL.Path, "/device/auth/verify_code")
+}
+
+func (s *Server) handleDeviceExchange(w http.ResponseWriter, r *http.Request) {
+	switch r.Method {
+	case http.MethodGet:
+		userCode := r.URL.Query().Get("user_code")
+		invalidAttempt, err := strconv.ParseBool(r.URL.Query().Get("invalid"))
+		if err != nil {
+			invalidAttempt = false
```

Comment:
I
![Screen Shot 2020-06-02 at 2 24 13 PM](https://user-images.githubusercontent.com/54368740/83555784-d023f800-a4dc-11ea-8c67-747a2e7f9a54.png)
f the user inputs an incorrect/invalid device code, this set a basic flag so that a user can be alerted on the page and try again.

Reply:
Maybe this is worth a comment to explain what invalidAttempt achieves? I think I'd forget if I came back to this in 6 months

Reply:
@JoelSpeed  fair enough, will do!